pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -200,8 +200,23 @@ async def consumer(
                 message_queue.task_done()
         except asyncio.CancelledError:
             log.info('Consumer worker cancelled. Gracefully terminating...')
-            for _ in range(max_tasks):
-                await semaphore.acquire()
+            # we can't know if last task was even scheduled,
```

Comment:
I wonder if we could solve this using the `locked()` method ([source](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Semaphore.locked)). ie. something like "acquire the first `n-1` locks, then check if the semaphore is still locked and if so acquire once more"?

Reply:
But we have no idea if this last lock will ever unlock. Odds are, it won't. We need to be cancelled precisely on line 200 to be able to acquire the last lock.

Reply:
What if we move the `await semaphore.acquire()` somewhere where it is the last thing awaited in the `while True` loop? Like just before we schedule the task:

```python
                await semaphore.acquire()
                task = asyncio.ensure_future(_execute_callback(
                    message,
                    callback,
                    ack_queue,
                    nack_queue,
                    metrics_client,
                ))
```

Reply:
But we still can be cancelled after acquiring a lock but before scheduling a task => deadlock

Reply:
Can we? AFAIK a task can only be cancelled at `await` statements (and I suppose before it begins) and not on any given line. Please enlighten me if I am mistaken! 

But I think `shield`ing as @jonathan-johnston pointed out would also work.

Reply:
Tbh I don't know how to read docs on that:
![image](https://user-images.githubusercontent.com/62574824/104506745-a6ff3600-559a-11eb-82c1-b4290247db1d.png)
![image](https://user-images.githubusercontent.com/62574824/104506792-b54d5200-559a-11eb-82df-dda4c8327149.png)
So to be cancelled a task needs a new cycle but if we're awaiting on a future, the future will get cancelled instead. I think you're right here.